plugins/filter_aws/aws.c

```diff
@@ -370,6 +385,134 @@ static int get_vpc_metadata(struct flb_filter_aws *ctx)
     return ret;
 }
 
+/* Get EC2 instance tag keys from /latest/meta-data/tags/instance.
+ * Initializes ctx->tags_count, ctx->tag_keys and ctx->tag_keys_len.
+ */
+static int get_ec2_tag_keys(struct flb_filter_aws *ctx)
+{
+    int ret;
+    flb_sds_t tags_list = NULL;
+    size_t len = 0;
+    size_t tag_index = 0;
+    size_t tag_start = 0;
+    size_t tag_end = 0;
+    size_t characters_to_copy;
+    size_t i;
+
+    /* get a list of tag keys from the meta data server */
+    ret = get_metadata(ctx, FLB_FILTER_AWS_IMDS_INSTANCE_TAG, &tags_list, &len);
+    if (ret < 0) {
+        if (ret == -2) {
+            flb_warn("[filter_aws] tags not available in the EC2 instance metadata");
+            ctx->tags_count = 0;
+            return 0;
+        }
+        flb_sds_destroy(tags_list);
+        return -1;
+    }
+    /* count number of tag keys and allocate memory */
+    /* it assumes there always is at least one tag */
+    /* starts with 1, because '\n' is a split character */
+    ctx->tags_count = 1;
+    for (i = 0; i < len; i++) {
+        if (tags_list[i] == '\n') {
+            ctx->tags_count++;
+        }
+    }
+    ctx->tag_keys = flb_calloc(ctx->tags_count, sizeof(flb_sds_t*));
+    ctx->tag_keys_len = flb_calloc(ctx->tags_count, sizeof(size_t*));
+
+    /* go over the response and initialize tag_keys values */
+    /* code below finds two indices which define tag key and copies them to ctx */
+    while (tag_end <= len) {
+        if (tags_list[tag_end] == '\n') {
+            tags_list[tag_end] = '\0';
+        }
+        if ((tags_list[tag_end] == '\0' || tag_end == len) && (tag_start < tag_end)) {
+            /* length of the key + 1 place for null character */
+            characters_to_copy = tag_end - tag_start;
+
+            ctx->tag_keys_len[tag_index] = characters_to_copy;
+            ctx->tag_keys[tag_index] = flb_sds_create_size(characters_to_copy + 1);
+            if (ctx->tag_keys[tag_index] == NULL) {
+                flb_sds_destroy(tags_list);
+                return -2;
+            }
+
+            flb_sds_t tag_key = tags_list + tag_start;
+            strcpy(ctx->tag_keys[tag_index], tag_key);
```

Comment:
I am confused by this code/to me it seems like it should not work. So the `ctx->tag_keys` is an array of pointers to an SDS string. This means you only allocated memory for the references, not the actual strings. 

So I think you need to do something like:

```
/* define all vars at top of func this is an example */
flb_sds_t tmp;
tmp = ctx->tag_keys[tag_index]; // get the reference
tmp = flb_sds_create(tag_key);
// now check for alloc failure
```

Basically, you must allocate new memory to hold these characters.

Reply:
First of all, I apologize for imperfect implementation which might be difficult to understand.

Okay, so let me try to explain what's going on. On a slightly higher level, what we do is:
 1. HTTP request for list of tags in the format `tagkey1\ntagkey2`, this is our `tags_list` which we are about to parse in the next steps;
    1.1. please note that at `tags_list` will have at least one item, otherwise HTTP response status code is 404 and we would exit earlier;
 2. Establish count of the items in the `tags_list`, our separator is `\n`, so count of the items is equal to the number of occurrences inside the `tags_list` + 1; this is our `ctx->tags_count`;
 3. allocate memory for ctx->tag_keys as an array of pointers,
 4. go over the tags_list, find pairs of indices which represent each tag key, and for each such pair:
   4.1. allocate memory for tag key (`ctx->tag_keys[tag_index] = flb_sds_create_size(characters_to_copy + 1);`),
   4.2. copy the tag key to the our newly allocated memory (`strcpy(ctx->tag_keys[tag_index], tag_key);`),

> So I think you need to do something like: [...] flb_sds_create(tag_key);

If I understand my code correctly, I do the same exact thing as you mentioned:

```c
            ctx->tag_keys[tag_index] = flb_sds_create_size(characters_to_copy + 1);
            // ...
            strcpy(ctx->tag_keys[tag_index], tag_key);
```

> Basically, you must allocate new memory to hold these characters.

Yes, I do, `flb_sds_create_size` allocates new memory for each tag key.

---

There is an alternative solution where we could just use the response bytes from the point 1., replace \n to \0 and populate the array of points such that each item points to the first character of each tag name. It wouldn't require allocation of new memory. It would be quicker to look up, because this way memory would be contagious so it would be able to use CPU cache more effectively. However, given the probably super low count of AWS tags (N<10), it's most likely negligible, whereas complexity of this solution would be even higher (so imo not worth it).

Reply:
I've made an attempt to improve readability of the implementation. https://github.com/fluent/fluent-bit/commit/b185ddf3aebbdc1d7b3a9a0661d42c88bacaf389

Let me know if you have any suggestions regarding the comments or the implementation itself. It might be the case there is a way easier way to achieve the same effect. (I haven't coded in C too much.)

Reply:
Cool/Thanks!. 

Can you also use `memcpy` here not `strcpy`. You can see examples in the existing code base. Eduardo (lead maintainer) prefers memcpy for these things, I've gotten comments in past PRs.

Reply:
Done, https://github.com/fluent/fluent-bit/pull/6065/commits/11e2d924d26273b1c6521fa8ed1d2cddc230eba8